csrc/id_model/loop_promotion.cpp

```diff
@@ -16,6 +16,184 @@
 
 namespace nvfuser {
 
+std::string toString(const CoveredGroups& covered_groups) {
+  std::stringstream ss;
+  ss << "{\n";
+  for (const auto& cg : covered_groups) {
+    ss << "\t" << cg.toString() << "\n";
+  }
+  ss << "}\n";
+  return ss.str();
+}
+
+// Check if this CoveredGroup is equal to or a superset of the other
+// CoveredGroup. Trying to check as many sufficient conditions as
+// possible, but may not be complete.
+bool CoveredGroup::isEqualToOrSuperSetOf(const CoveredGroup& other) const {
+  if (*this == other) {
+    return true;
+  }
+
+  // When both are derived from split and correspond to either inner
+  // or outer.
+  if (split_in_.get() && other.split_in_.get() &&
+      is_inner_ == other.is_inner_) {
+    const CoveredGroups& split_in = *split_in_;
+    const CoveredGroups& other_split_in = *other.split_in_;
+
+    // When both have the same split input (and both correspond to
+    // either inner or outer), they should cover the same exact
+    // groups. This should only happen when broadcast is merged. For
+    // example, suppose there are two tensors and they are scheduled as
+    // follows;
+    //
+    // t0: [i0]
+    // t1: [i1, b2]
+    //
+    // t1->merge(0, 1)->split(0, 4);
+    // t0->split(0, 4)
+    //
+    // t0->inlineAt(t1, 1)
+    //
+    // In this case, t0->axis(0) and t1->axis(0) have the same
+    // split input group, {i0, i1}. Note that b2 is not included as
+    // it's a broadcast. Also note that both are the outer
+    // output. Here, group_ of t0->axis(0) is the exact group of
+    // t0->axis(0), while that of tv1->axis(0) is the exact group of
+    // the t1->merge(0, 1) output. In this case, however, this merge
+    // is just a merge of i1 and the b2 broadcast ID, so in terms of
+    // covered exact groups, it's effectively the same as that of
+    // t0->axis(0). All in all, as long as both correspond to either
+    // inner or outer of the same split input, they should be
+    // considered the same.
+    if (split_in == other_split_in) {
+      return true;
+    }
+
+    // Both are derived from a split but have differnt split input
+    // groups. If the input groups of this split is a superset of the
+    // input groups of the split of the other CoveredGroup, this
+    // CoveredGroup is a superset
+    if (std::all_of(
+            other_split_in.begin(),
+            other_split_in.end(),
+            [&](const CoveredGroup& other_split_in_group) {
+              return std::any_of(
+                  split_in.begin(),
+                  split_in.end(),
+                  [&](const CoveredGroup& split_in_group) {
+                    return split_in_group.isEqualToOrSuperSetOf(
+                        other_split_in_group);
+                  });
+            })) {
```

Comment:
Can this  be simplified as the following?

```C++
if (isEqualToOrSuperSetOf(*split_in, *other_split_in)) {
```

Reply:
Yes, thanks for pointing it out. Will update.

Reply:
Updated